### PR TITLE
fix: only outdated generators should be updated

### PR DESCRIPTION
**Pull request by slavik-lvovsky**

### CODE REVIEW

File: packages/backend/src/utils/env.ts [hunk 1]
```diff
@@ -92,39 +90,52 @@ class EnvUtil {
...
   }
 
-  public getGenNamespaces(): string[] {
-    const gensMeta = this.getGeneratorsMeta(false);
+  private getGensMetaByInstallationPath(): Environment.LookupGeneratorMeta[] {
+    const npmInstallationPaths = [customLocation.getNodeModulesPath() ?? NpmCommand.getGlobalNodeModulesPath()];
+    return this.lookupGensMeta({ npmPaths: npmInstallationPaths });
+  }
+
+  public getAllGeneratorNamespaces(): string[] {
```
Comment (alex-gilin) [inline packages/backend/src/utils/env.ts:102]:
```suggestion
  public getAllGeneratorNamespaces(): string[] | undefined {
```

File: packages/backend/src/utils/env.ts [hunk 2]
```diff
@@ -92,39 +90,52 @@ class EnvUtil {
...
+    const gensMeta: Environment.LookupGeneratorMeta[] = this.getGeneratorsMeta(false);
     return _.map(gensMeta, (genMeta) => genMeta.namespace);
   }
 
   public createEnvAndGen(genNamespace: string, options: any, adapter: any): any {
```
Comment (alex-gilin) [inline packages/backend/src/utils/env.ts:107]:
```suggestion
  public createEnvAndGen(genNamespace: string, options: any, adapter: any): { env: Environment<Environment.Options> ; gen: any>} {
```

File: packages/backend/src/utils/npm.ts [hunk 3]
```diff
@@ -120,6 +109,50 @@ class Command {
     }
   }
 
+  private async shouldBeUpdated(packageJson: any): Promise<boolean> {
+    const queryUrl = this.getSingleGenQueryURL(packageJson.name);
+    const npmJsModules = await npmFetch.json(queryUrl);
+    const npmJsModule: any = _.get(npmJsModules, "objects.[0]");
+    return npmJsModule ? npmJsModule.package.version !== packageJson.version : false;
```
Comment (alex-gilin) [inline packages/backend/src/utils/npm.ts:116]:
```suggestion
    return npmJsModule ? npmJsModule.package?.version !== packageJson.version : false;
```

File: packages/backend/src/utils/npm.ts [hunk 4]
```diff
@@ -120,6 +109,50 @@ class Command {
...
+  }
+
+  public getGlobalNodeModulesPath(): string {
+    return this.globalNodeModulesPath;
+  }
+
+  public async getPackagesMetadata(query = "", author = "") {
```
Comment (alex-gilin) [inline packages/backend/src/utils/npm.ts:123]:
```suggestion
  public async getPackagesMetadata(query = "", author = ""): Promise<any> {
```

File: packages/backend/src/utils/npm.ts [hunk 5]
```diff
@@ -120,6 +109,50 @@ class Command {
...
+    const gensQueryUrl = NpmCommand.getGensQueryURL(query, author);
+    return await npmFetch.json(gensQueryUrl);
+  }
+
+  public async getPackageJson(packagePath: string): Promise<any> {
+    const packageJsonString: string = await readFile(path.join(packagePath, "package.json"), "utf8");
+    return JSON.parse(packageJsonString);
+  }
```
Comment (alex-gilin) [inline packages/backend/src/utils/npm.ts:131]: seems like unsafe to use -> please consider to add error handling

File: packages/backend/src/utils/npm.ts [hunk 6]
```diff
@@ -120,6 +109,50 @@ class Command {
...
+
+  public async getPackageNamesWithOutdatedVersion(packagePaths: string[]): Promise<string[]> {
+    const packageJsonPromises: any[] = packagePaths.map((packageJsonPath) => this.getPackageJson(packageJsonPath));
+    const packageJsons: any[] = await Promise.all(packageJsonPromises);
```
Comment (alex-gilin) [inline packages/backend/src/utils/npm.ts:135]: what you intend here in case one/any promises rejected ?

**Review by alex-gilin — CHANGES_REQUESTED** [review]

File: packages/backend/src/exploregens.ts [hunk 7]
```diff
@@ -101,11 +100,11 @@ export class ExploreGens {
...
   private async updateAllInstalledGenerators() {
-    const installedGenerators: string[] = this.getAllInstalledGenerators();
-    if (!_.isEmpty(installedGenerators)) {
+    const gensToUpdate: string[] = await Env.getGenNamesWithOutdatedVersion();
+    if (!_.isEmpty(gensToUpdate)) {
       this.logger.debug(messages.auto_update_started);
       const statusBarMessage = vscode.window.setStatusBarMessage(messages.auto_update_started);
-      const promises = _.map(installedGenerators, (genName) => this.update(genName));
+      const promises = _.map(gensToUpdate, (genName) => this.update(genName));
       const failedToUpdateGens: any[] = _.compact(await Promise.all(promises));
```
[the remark this answers is not in the document]
Comment (slavik-lvovsky) [reply packages/backend/src/exploregens.ts:112]: try/catch is added

File: packages/backend/src/utils/npm.ts (hunk 3 above)
Comment (slavik-lvovsky) [reply packages/backend/src/utils/npm.ts:118]: it is always exists here

File: packages/backend/src/utils/npm.ts (hunk 4 above)
Comment (slavik-lvovsky) [reply packages/backend/src/utils/npm.ts:135]: it is the same in this case

File: packages/backend/src/utils/npm.ts (hunk 5 above)
Comment (slavik-lvovsky) [reply packages/backend/src/utils/npm.ts]: added try/catch

File: packages/backend/src/utils/npm.ts (hunk 6 above)
Comment (slavik-lvovsky) [reply packages/backend/src/utils/npm.ts]: an exception isnow handled in getPackageJson with try/catch

File: packages/backend/src/utils/env.ts [hunk 8]
```diff
@@ -22,15 +23,15 @@ class EnvUtil {
...
     // improves lookup speed by 1.5 sec
     this.existingNpmPaths = (() => {
-      const globalNpmPaths = this.createEnvInstance().getNpmPaths();
+      const globalNpmPaths: string[] = this.createEnvInstance().getNpmPaths();
       const userNpmPaths = homedir()
         .split(path.sep)
         .map((part, index, parts) => {
           const resPath = path.join(...parts.slice(0, index + 1), "node_modules");
           return isWin32 ? resPath : path.join(path.sep, resPath);
         });
```
[the remark this answers is not in the document]
Comment (slavik-lvovsky) [reply packages/backend/src/utils/env.ts:37]: what it should improve ?

File: packages/backend/src/utils/env.ts (hunk 1 above)
Comment (slavik-lvovsky) [reply packages/backend/src/utils/env.ts:112]: I use map, so it always be an array

File: packages/backend/src/exploregens.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (alex-gilin) [reply packages/backend/src/exploregens.ts:112]: if you want to continue to wait until the rest of promises will completed ( but not be rejected by first exception on any of promises) you should use other logic here

File: packages/backend/src/utils/env.ts (hunk 2 above)
Comment (slavik-lvovsky) [reply packages/backend/src/utils/env.ts]: Added new type

File: packages/backend/src/utils/env.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (alex-gilin) [reply packages/backend/src/utils/env.ts:42]: creation nearer to usage

File: packages/backend/src/utils/npm.ts (hunk 3 above)
Comment (alex-gilin) [reply packages/backend/src/utils/npm.ts:118]: but it is not covered by type definition. seems like can get exception in runtime

File: packages/backend/src/utils/npm.ts (hunk 4 above)
Comment (alex-gilin) [reply packages/backend/src/utils/npm.ts:135]: I know, but suitable eslint rule seems like turned off and the other apis have returns type in function signature

File: packages/backend/src/utils/npm.ts (hunk 5 above)
Comment (alex-gilin) [reply packages/backend/src/utils/npm.ts]: like above

**Review by alex-gilin — APPROVED** [review]

**Comment by slavik-lvovsky** [comment]
Validated successfully on BAS

**Pull request merged**